include/aklite.h

```diff
@@ -0,0 +1,178 @@
+  };
+  DownloadResult(Status status, std::string description = "") : status(status), description(std::move(description)) {}
+  Status status;
+  std::string description;
+};
+
+/**
+ * The response from an AkliteClient call to Install
+ */
+class InstallResult {
+ public:
+  enum class Status {
+    Ok,
+    NeedsCompletion,
+    Failed,
+  };
+  InstallResult(Status status, std::string description = "") : status(status), description(std::move(description)) {}
+  Status status;
+  std::string description;
+};
+
+/**
+ * AkliteClient provides an easy-to-use API for users wanting to customize
+ * the behavior of aktualiz-lite.
+ */
+class AkliteClient {
+ public:
+  /**
+   * Construct a client instance pulling in config files from the given
+   * locations. ex:
+   *
+   *   AkliteClient c(AkliteClient::CONFIG_DIRS)
+   *
+   * @param config_dirs The list of files/directories to parse sota toml from.
+   */
+  AkliteClient(const std::vector<boost::filesystem::path> &config_dirs);
+  /**
+   * Used for unit-testing purposes.
+   */
+  AkliteClient(std::shared_ptr<LiteClient> client) : client_(client) {}
+
+  /**
+   * Performs a "check-in" with the device-gateay. A check-in consists of:
+   *  1) ask device-gateway for a new root.json - normally a 404.
+   *  2) ask device-gateway for timestamp and snapshot metadata.
+   *  3) pull down a new targets.json if needed
+   */
+  CheckInResult CheckIn(std::string hwid = "") const;
+
+  /**
+   * Return the active aktualizr-lite configuration.
+   */
+  boost::property_tree::ptree GetConfig() const;
+
+  /**
+   * Report network and hwinfo to server. This only reports data if the
+   * information has changed locally.
+   */
+  void SendTelemetry() const;
+
+  /**
+   * Report the full aktualizr-lite active configuration to the
+   * device-gateway. Should be called once during start-up .
+   */
+  void ReportAkliteConfiguration() const;
+
+  /**
+   * Return the Target currently running on the system.
+   */
+  TufTarget GetCurrent() const;
+
+  /**
+   * Download and verify the content of the given target.
+   */
+  DownloadResult Download(const TufTarget &t, const std::string &correlation_id, std::string reason = "");
+
+  /**
+   * Install the Target
+   */
+  InstallResult Install(const TufTarget &t, const std::string &correlation_id);
+
+  /**
+   * Check if the Target has been installed but failed to boot. This would
+   * make this be considered a "rollback target" and one we shouldn't consider
+   * installing.
+   */
+  bool IsRollback(const TufTarget &t) const;
+
+  /**
+   * Default files/paths to search for sota toml when configuration client.
+   */
+  std::vector<boost::filesystem::path> CONFIG_DIRS = {"/usr/lib/sota/conf.d", "/var/sota/sota.toml",
+                                                      "/etc/sota/conf.d/"};
+
+  /**
+   * Return a random UUID to be used for a Download/Install operation
+   */
+  static std::string GenerateCorrelationId(const TufTarget &t);
```

Comment:
The most common source of friction and bugs I get when using this code base has always been how stateful objects are without you knowing. You get code where you have to know to call methodA before methodB is safe. Or having to remember to call our magic method https://github.com/foundriesio/aktualizr-lite/blob/master/src/main.cc#L325.

Setting that in TufTarget is going to lead to it getting set internally, the update working, the user doing `current = latest` and then the next update won't get a good correlation id.

One alternative I was thinking is having the DownloadResponse include a method `Install`. Then it could all be encapsulated there. The problem I'm seeing is that the main reason people want to customize our agent is because they have very unique ways they want to control the update. e.g. maybe download, wait for some time - perhaps the device or update agent would restart. and then install. I'm not sure this compromise would help them. But maybe I'm dreaming complications that don't exist - nobody has specifically asked for that.

Reply:
What do you all think of something like this change:

 59726560c529483fe5288bb7f277700d4d614a08

Its kind of interesting. We really don't want people to call Install if Download hasn't been called, so this kind of forces things to be done correctly.

Reply:
That's a pretty cool idea as far as I'm concerned.